### DISCUSSION TITLE
TypeError: __init__() missing 4 required positional arguments: 'consumer_key', 'consumer_secret', 'access_token', and 'access_token_secret'

### DISCUSSION THREAD

**Discussion by interzone2** [tweepy/tweepy#1681] [Questions] [answered]
On local dev machine (OS X 11.5.1) this code is working [Python 3.8]. The second version run on an ubuntu 20.04.3 VPS is not working throwing the error. 

```
from tweepy import Stream
from tweepy import OAuthHandler
from tweepy.streaming import StreamListener
import json
import sqlite3
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from unidecode import unidecode

analyzer = SentimentIntensityAnalyzer()

#consumer key, consumer secret, access token, access secret.
ckey="????"
csecret="????"
atoken="????"
asecret="????"

conn = sqlite3.connect('twitter.db')
c = conn.cursor()

def create_table():
    try:
        c.execute("CREATE TABLE IF NOT EXISTS sentiment(unix REAL, tweet TEXT, sentiment REAL)")
        c.execute("CREATE INDEX fast_unix ON sentiment(unix)")
        c.execute("CREATE INDEX fast_tweet ON sentiment(tweet)")
        c.execute("CREATE INDEX fast_sentiment ON sentiment(sentiment)")
        conn.commit()
    except Exception as e:
        print(str(e))
create_table()

class listener(StreamListener):

    def on_data(self, data):
        try:
            data = json.loads(data)
            tweet = unidecode(data['text'])
            time_ms = data['timestamp_ms']
            vs = analyzer.polarity_scores(tweet)
            sentiment = vs['compound']
            print(time_ms, tweet, sentiment)
            c.execute("INSERT INTO sentiment (unix, tweet, sentiment) VALUES (?, ?, ?)",
                  (time_ms, tweet, sentiment))
            conn.commit()

        except KeyError as e:
            print(str(e))
        return(True)

    def on_error(self, status):
        print(status)

while True:

    try:
        auth = OAuthHandler(ckey, csecret)
        auth.set_access_token(atoken, asecret)
        twitterStream = Stream(auth, listener())
        twitterStream.filter(track=["a","e","i","o","u"])
    except Exception as e:
        print(str(e))
        time.sleep(5)
```

On my **Ubuntu VPS 20.04.3 Python 3.8** server I have changed the first import to the second below as it was causing error related to updates in tweepy. 

`from tweepy.streaming import StreamListener` 

`from tweepy.streaming import Stream` 

But keep getting the error below, even though nothing much is different in the setups in either location. Except: Using Flask and Dash on Ubuntu. Running this test here: **/var/www/FlaskApp/FlaskApp/misc**

I assume i need to deal with AUTH in a different way?

```
Traceback (most recent call last):
  File "car2.py", line 52, in <module>
    twitterStream = Stream(auth, listener())
TypeError: __init__() missing 4 required positional arguments: 'consumer_key', 'consumer_secret', 'access_token', and 'access_token_secret'
```

**Comment by Harmon758** [comment, marked as answer]
[Tweepy v4.0.0](https://github.com/tweepy/tweepy/releases/tag/v4.0.0) merged `StreamListener` into [`Stream`](https://tweepy.readthedocs.io/en/v4.0.1/stream.html#tweepy.Stream) and changed `Stream` to accept individual `consumer_key`, `consumer_secret`, `access_token`, and `access_token_secret` parameters rather than an `auth` parameter.
You're using a newer version of Tweepy on your VPS.

**Reply by daradeepak** [reply]
??

**Reply by daradeepak** [reply]
How to make stream accept individual consumer_key, consumer_secret, access_token, and access_token_secret parameters rather than an auth parameter

**Reply by Harmon758** [reply]
By using Tweepy v4

**Reply by popcorn5372** [reply]
Ok but what is the fix for it? Many sets of example code use this line and now it's all broken, and people are google for answers and end up on this particular thread - that doesn't have a fix.

**Reply by Harmon758** [reply]
This isn't a bug with Tweepy.
You can fix your code by updating for the corresponding changes to `Stream` in Tweepy v4.
For this specific error, that would be passing the mentioned arguments/parameters.
Alternatively, you can downgrade to v3.10.0.

**Reply by popcorn5372** [reply]
The issue here is that there isn't an easily accessible guide for making the corresponding changes, and this breaks things for people at beginner/intro level.

For anyone who stumbles upon this page - Fix in command line to downgrade is:

> pip3 uninstall tweepy
> pip3 install -Iv tweepy==3.10.0

This will make original code referenced across youtube work

**Reply by Harmon758** [reply]
[The reference documentation for `Stream`](https://tweepy.readthedocs.io/en/v4.2.0/stream.html) and [the page on streaming in the documentation](https://tweepy.readthedocs.io/en/v4.2.0/streaming.html) cover usage of `Stream`, and the [changelog](https://tweepy.readthedocs.io/en/v4.2.0/changelog.html#version-4-0-0) and [release notes](https://github.com/tweepy/tweepy/releases/tag/v4.0.0) for Tweepy v4.0.0 cover the specific changes fairly comprehensively.

In terms of Tweepy's documentation for `Stream` and streaming, v4 actually is much more complete, with the reference documentation only having been added with v4.0.0.

I would recommend against downgrading to v3.10.0. There were a great deal of improvements and fixes in v4.0.0, including to streaming, (see the changelog / release notes) and if anything, the class interface should now be more intuitive, to beginners and non-beginners alike. Additionally, v3 will not have any more releases and has many known bugs/issues, including with streaming, that were fixed in v4.0.0, that will never be fixed for v3, and any new bugs or breakage due to Twitter API changes will not be fixed for v3. v3 is no longer supported, and v4 will be the version that is maintained, updated, and improved going forward.

Also, there's no reason to use the `--ignore-installed` flag if you already uninstalled. (It's more likely to cause issues if anything.) In fact, there's no reason to uninstall first. pip will install the specified version even if it's a downgrade.

It is not up to libraries and library maintainers to provide indefinitely backward compatible code to "make original code referenced across youtube work". There will be breaking changes as code is improved, and the major version bump from v3 to v4 indicates that that's the case here. The relevant Tweepy documentation was updated prior to release, but it is up to the writers of third-party guides to update accordingly. Personally, this is one of the reasons I recommend against using YouTube videos for something like learning how to use a library; they can often be out of date since they aren't easily updateable by their creators, but to each their own.

**Reply by JoIyke** [reply]
> How to make stream accept individual consumer_key, consumer_secret, access_token, and access_token_secret parameters rather than an auth parameter

You can check to see if my solution works for you. It worked for me.

**Comment by y3khan** [comment]
hi everyone, this solution doesn't really. any sample code here?

**Reply by Harmon758** [reply]
I'm not sure what you're saying, but you can find [a basic example of streaming in the documentation](https://tweepy.readthedocs.io/en/v4.4.0/examples.html?tab=Streaming).

**Comment by JoIyke** [comment]
The solution above did not work for me, maybe because it is not clear enough for beginners. 

You can replace the entire `while True` block at the end with this:

```
twitterStream = listener(ckey, csecret, atoken, asecret)
twitterStream.filter(track=["a","e","i","o","u"])
```

I also had to call:
`def listener(tweepy.Stream):
     ...
`
instead of 
`def listener(StreamListener):
`

I hope this helps.
